pkg/parse/root.go

```diff
@@ -334,55 +338,193 @@ func (p *root) setRenderingStatusWithRetires(ctx context.Context, newStatus rend
 	return nil
 }
 
-func setRenderingStatusFields(rendering *v1beta1.RenderingStatus, p Parser, newStatus renderingStatus, denominator int) {
-	cse := status.ToCSE(newStatus.errs)
-	rendering.Commit = newStatus.commit
-	switch p.options().SourceType {
+func setRenderingStatusFields(renderingStatus *v1beta1.RenderingStatus, newStatus RenderingStatus, denominator int) {
+	cse := status.ToCSE(newStatus.Errs)
+	renderingStatus.Commit = newStatus.Commit
+	switch newSourceSpec := newStatus.Spec.(type) {
+	case GitSourceSpecStatus:
+		renderingStatus.Git = &v1beta1.GitStatus{
+			Repo:     newSourceSpec.Repo,
+			Revision: newSourceSpec.Revision,
+			Branch:   newSourceSpec.Branch,
+			Dir:      newSourceSpec.Dir,
+		}
+		renderingStatus.Oci = nil
+		renderingStatus.Helm = nil
+	case OCISourceSpecStatus:
+		renderingStatus.Oci = &v1beta1.OciStatus{
+			Image: newSourceSpec.Image,
+			Dir:   newSourceSpec.Dir,
+		}
+		renderingStatus.Git = nil
+		renderingStatus.Helm = nil
+	case HelmSourceSpecStatus:
+		renderingStatus.Helm = &v1beta1.HelmStatus{
+			Repo:    newSourceSpec.Repo,
+			Chart:   newSourceSpec.Chart,
+			Version: newSourceSpec.Version,
+		}
+		renderingStatus.Git = nil
+		renderingStatus.Oci = nil
+	default:
+		renderingStatus.Helm = nil
+		renderingStatus.Git = nil
+		renderingStatus.Oci = nil
+	}
+	renderingStatus.Message = newStatus.Message
+	errorSummary := &v1beta1.ErrorSummary{
+		TotalCount: len(cse),
+		Truncated:  denominator != 1,
+	}
+	renderingStatus.Errors = cse[0 : len(cse)/denominator]
+	renderingStatus.ErrorSummary = errorSummary
+	renderingStatus.LastUpdate = newStatus.LastUpdate
+}
+
+// SyncStatus implements the Parser interface
+// SyncStatus gets the RootSync sync status.
+func (p *root) LoadStatus(ctx context.Context) (*ReconcilerStatus, error) {
+	p.mux.Lock()
+	defer p.mux.Unlock()
+
+	rs := &v1beta1.RootSync{}
+	if err := p.Client.Get(ctx, rootsync.ObjectKey(p.SyncName), rs); err != nil {
+		if apierrors.IsNotFound(err) || meta.IsNoMatchError(err) {
+			return nil, nil
+		}
+		return nil, status.APIServerError(err, "failed to get RootSync")
+	}
+
+	syncing := false
+	var syncingConditionLastUpdate metav1.Time
+	for _, condition := range rs.Status.Conditions {
+		if condition.Type == v1beta1.RootSyncSyncing {
+			if condition.Status == metav1.ConditionTrue {
+				syncing = true
+			}
+			syncingConditionLastUpdate = condition.LastUpdateTime
+			break
+		}
+	}
+
+	return reconcilerStatusFromRSyncStatus(rs.Status.Status, p.options().SourceType, syncing, syncingConditionLastUpdate), nil
+}
+
+func reconcilerStatusFromRSyncStatus(rsyncStatus v1beta1.Status, sourceType v1beta1.SourceType, syncing bool, syncingConditionLastUpdate metav1.Time) *ReconcilerStatus {
```

Comment:
This func is particularly repetitive. But the only way I can think to fix it would just be to not use an internal representation at all and just use the v1beta1 structs. However, mixing external and internal structs for the in-memory cache seems messy. Ideally it would just cache the whole RSync internally, or have some internal type that can easily map to both RepoSync & RootSync.

If I was doing this all over again, I would just make RepoSync & RootSync the same CRD. That would save so much headache, tech debt, and code duplication.